### PR TITLE
Refactor the prepare connector to a separate method

### PR DESCRIPTION
**Pull request by wangch079**
This PR did a small refactoring, to extract the prepare connector logic into a separate method.

#### Pre-Review Checklist
- [ ] Covered the changes with automated tests
- [x] Tested the changes locally
- [x] Added a label for each target release version (example: `v7.13.2`, `v7.14.0`, `v8.0.0`)
- [ ] Considered corresponding documentation changes
- [ ] Contributed any configuration settings changes to the configuration reference

### CODE REVIEW

File: connectors/services/sync.py [hunk 1]
```diff
@@ -164,3 +148,27 @@ async def _run(self):
                 await self.sync_job_index.close()
             await es.close()
         return 0
+
+    async def _prepare_connector(self, connector):
```
Comment (artem-shelkovnikov) [inline connectors/services/sync.py:152]: What do you think of adding a docstring for this method @wangch079?

**Review by artem-shelkovnikov — APPROVED** [review]
LGTM, added a suggestion about docstring

File: connectors/services/sync.py (hunk 1 above)
Comment (wangch079) [reply connectors/services/sync.py:152]: This method `_prepare_connector` just calls `connector.prepare()`, catches all the possible exceptions, and then returns `True` or `False` indicating whether this operation is successful.

I'm not sure whether I should add `docstring` to a **private** method, so I updated the `docstring` of `Connector.prepare()` method.

File: connectors/byoc.py [hunk 2]
```diff
@@ -586,9 +586,11 @@ async def prepare_docs(self, data_provider, pipeline=None, filtering=None):
             yield doc, lazy_download
 
     async def prepare(self, config):
-        """Prepares the connector, given a configuration
-        If the connector id and the service type is in the config, we want to
-        populate the service type and then sets the default configuration.
+        """
```
Comment (tarekziade) [inline connectors/byoc.py:589]: pep8, don't leave an empty line when you start docstrings

**Review by tarekziade — COMMENTED** [review]
I am wondering what is the goal of this refactoring.

**Comment by wangch079** [comment]
>I am wondering what is the goal of this refactoring.

To make the `_one_sync` function more readable. Actually what the method `_prepare_connector` does is to call `connector.prepare`, and catch all the exceptions, log the error, and return `True` or `False`. 

I think we don't even need `_prepare_connector` method. We can just log the errors inside `connector.prepare`. WDYT?

**Comment by tarekziade** [comment]
> > I am wondering what is the goal of this refactoring.
> 
> To make the `_one_sync` function more readable. Actually what the method `_prepare_connector` does is to call `connector.prepare`, and catch all the exceptions, log the error, and return `True` or `False`.
> 
> I think we don't even need `_prepare_connector` method. 

I don't find it more readable because you are adding more lines, with some `return True` and `return False` and creating a new method that is only used by `_one_sync` - and introducing a change where you catch all exceptions.

The original code is smaller and easy to read in one place. 

> We can just log the errors inside `connector.prepare`. WDYT?

That is a different change, `connector.prepare` was built by raising errors and it's already logging them.

If your goal is to reduce the size of `_one_sync`, maybe the trick is to have only one try..catch:

```python
        try:
            await connector.prepare(self.config)
        except ConnectorPrepareError as e:
            if e.critical:
                await connector.error(e)
                logger.critical(e, exc_info=True)
                self.raise_if_spurious(e)
            else:
                logger.error(e.msg)
            return
```

and create a `ConnectorPrepareError` with a hierarchy of errors, and a `critical` flag

**Comment by wangch079** [comment]
>The original code is smaller and easy to read in one place.

👌 

> If your goal is to reduce the size of `_one_sync`, maybe the trick is to have only one try..catch:
> and create a `ConnectorPrepareError` with a hierarchy of errors, and a `critical` flag

I think it's better to log different messages for different errors. 

Will close this PR.

**Comment by tarekziade** [comment]
> I think it's better to log different messages for different errors.

this is what a hierarchy of errors gives you:

```
class ConnectorPrepareError(Exception):
    pass

class ServiceTypeNotConfiguredError(ConnectorPrepareError):
    msg = 'Service not configured'
    critical = False

```

**Pull request closed without merging**